src/range_map.rs

```diff
@@ -6,137 +6,112 @@
+        let len = len.bytes();
+        // Compute a slice starting with the elements we care about
+        let slice: &[Elem<T>] = if len == 0 {
+                // We just need any empty iterator.  We don't even want to
+                // yield the element that surrounds this position.
+                &[]
+            } else {
+                let first_idx = self.find_offset(offset);
+                &self.v[first_idx..]
+            };
+        let end = offset + len; // the first offset that is not included any more
+        slice.iter()
+            .take_while(move |elem| elem.range.start < end)
+            .map(|elem| &elem.data)
     }
 
     pub fn iter_mut_all<'a>(&'a mut self) -> impl Iterator<Item = &'a mut T> + 'a {
-        self.map.values_mut()
+        self.v.iter_mut().map(|elem| &mut elem.data)
     }
 
-    fn split_entry_at(&mut self, offset: u64)
+    // Split the element situated at the given `index`, such that the 2nd one starts at offset `split_offset`.
+    // Do nothing if the element already starts there.
+    // Return whether a split was necessary.
+    fn split_index(&mut self, index: usize, split_offset: u64) -> bool
     where
         T: Clone,
     {
-        let range = match self.iter_with_range(offset, 1).next() {
-            Some((&range, _)) => range,
-            None => return,
-        };
-        assert!(
-            range.start <= offset && range.end > offset,
-            "We got a range that doesn't even contain what we asked for."
-        );
-        // There is an entry overlapping this position, see if we have to split it
-        if range.start < offset {
-            let data = self.map.remove(&range).unwrap();
-            let old = self.map.insert(
-                Range {
-                    start: range.start,
-                    end: offset,
-                },
-                data.clone(),
-            );
-            assert!(old.is_none());
-            let old = self.map.insert(
-                Range {
-                    start: offset,
-                    end: range.end,
-                },
-                data,
-            );
-            assert!(old.is_none());
+        let elem = &mut self.v[index];
+        if split_offset == elem.range.start || split_offset == elem.range.end {
+            // Nothing to do
+            return false;
         }
+        debug_assert!(elem.range.contains(&split_offset),
+            "The split_offset is not in the element to be split");
+
+        // Now we really have to split.  Reduce length of first element.
+        let second_range = split_offset..elem.range.end;
+        elem.range.end = split_offset;
+        // Copy the data, and insert 2nd element
+        let second = Elem {
+            range: second_range,
+            data: elem.data.clone(),
+        };
+        self.v.insert(index+1, second);
+        return true;
     }
 
     /// Provide mutable iteration over everything in the given range.  As a side-effect,
```

Comment:
a further side effect not mentioned in the docs is that adjoining elements are merged